Rover/Map.cpp

```diff
@@ -0,0 +1,327 @@
+    float len = sqrt(pq.first * pq.first + pq.second * pq.second);
+    pq.first = length * pq.first / len;
+    pq.second = length * pq.second / len;
+
+    point p1 = std::make_pair(q.first + pq.first, q.second + pq.second);
+    point p2 = std::make_pair(p.first - pq.first, p.second - pq.second);
+    return(std::make_pair(p1, p2));   
+}
+
+RoverPathfinding::point RoverPathfinding::Map::center_point_with_radius(RoverPathfinding::point cur, RoverPathfinding::point p, RoverPathfinding::point q, float R)
+{
+    point vec = std::make_pair(-p.second + q.second, p.first - q.first);
+    float len = sqrt((vec.first * vec.first) + (vec.second * vec.second));
+    vec.first = vec.first * R / len;
+    vec.second = vec.second * R / len;
+    point result = std::make_pair((p.first + q.first) / 2.0f, (p.second + q.second) / 2.0f);
+    int o = orientation(cur, p, q);
+    if(o == 1)
+    {
+	result.first += vec.first;
+	result.second += vec.second;
+    }
+    else //o is 2
+    {
+	result.first -= vec.first;
+	result.second -= vec.second;
+    }
+    return(result);
+}
+
+float RoverPathfinding::Map::dist_sq(point p1, point p2)
+{
+    return((p1.first - p2.first) * (p1.first - p2.first) + (p1.second - p2.second) * (p1.second - p2.second));
+}
+
+bool RoverPathfinding::Map::within_radius(point p1, point p2, float R)
+{
+    return(dist_sq(p1, p2) <= R * R);
+}
+
+void RoverPathfinding::Map::add_edge(int n1, int n2)
+{
+    float dist = sqrt(dist_sq(nodes[n1].coord, nodes[n2].coord));
+    
+    auto n1_to_n2 = std::make_pair(n2, dist);
+    nodes[n1].connection.push_back(n1_to_n2);
+
+    auto n2_to_n1 = std::make_pair(n1, dist);
+    nodes[n2].connection.push_back(n2_to_n1);   
+}
+
+int RoverPathfinding::Map::create_node(point coord)
+{
+    node n;
+    n.prev = -1;
+    n.dist_to = INFINITY;
+    n.coord = coord;
+    nodes.push_back(n);
+    return(nodes.size() - 1);
+}
+
+std::vector<RoverPathfinding::node> RoverPathfinding::Map::build_graph(point cur, point tar)
+{
+    std::queue<int> q;
+
+    //TODO(sasha): make R a constant - the following few lines are just a hack
+    //             to get R to be in lat/lng units
+#define R_METERS 0.5f
+    auto offset = lat_long_offset(cur.first, cur.second, 0.0f, R_METERS);
+    auto diff = std::make_pair(offset.first - cur.first, offset.second - cur.second);
+    float R = sqrt(diff.first * diff.first + diff.second * diff.second);
+#undef R_METERS
+    //</hack>		
+    
+    node start;
+    for(auto &n : nodes)
+    {
+	n.dist_to = INFINITY;
+	n.connection.clear();
+    }
+
+    nodes[0].prev = -1;
+    nodes[0].dist_to = 0.0f;
+    nodes[0].coord = cur;
+
+    nodes[1].coord = tar;
+
+    if(obstacles.empty())
+    {
+	add_edge(0, 1);
+	return(nodes);
+    }
+
+    q.push(0);
+    while(!q.empty())
+    {
+	int curr_node = q.front();
+	q.pop();
+	bool destination_blocked = false;
+	for(auto &obst : obstacles)
+	{
+	    if(segments_intersect(nodes[curr_node].coord, obst.coord1, tar, obst.coord2))
```

Comment:
It might be a good idea to extract this out into a couple functions to keep everything at the same level of abstraction